### PR TITLE
Enable qos parameter overrides for the /parameter_events topic 

### PR DESCRIPTION
**Pull request by ivanpauno**
This is a bit complex because the final qos profile is needed before a parameters node interface is created.

For `/rosout` it will even be more complex, because the final profile is needed even before a `rcl_node_t` is written!! (it's possible, but it requires some surgery probably involving `rcl`).
I'm not sure what's the best way to handle that one, opinions welcomed!

### CODE REVIEW

**Comment by wjwwood** [comment]
> For /rosout it will even be more complex, because the final profile is needed even before a rcl_node_t is written!! (it's possible, but it requires some surgery probably involving rcl).
I'm not sure what's the best way to handle that one, opinions welcomed!

I would say anything that needs configuration before parameters are setup (like rosout) need to have their own special handling in the command line arguments just like remapping arguments and log level.

File: rclcpp/src/rclcpp/node.cpp [hunk 1]
```diff
@@ -126,7 +170,7 @@ Node::Node(
       options.parameter_overrides(),
       options.start_parameter_services(),
       options.start_parameter_event_publisher(),
-      options.parameter_event_qos(),
+      get_parameter_events_qos(*node_base_, options),
```
Comment (wjwwood) [inline rclcpp/src/rclcpp/node.cpp:173]: It would be nice, either here or in the documentation of this function being called to indicate what we're doing (getting the QoS overrides from the parameter overrides directly) and why (because we need them before parameters can be bootstrapped).

File: rclcpp/test/rclcpp/test_node.cpp [hunk 2]
```diff
@@ -1966,14 +1966,14 @@ TEST_F(TestNode, get_parameters_undeclared_parameters_not_allowed) {
...
-    node_local->declare_parameter(name2, 100);
+    node_local->declare_parameter("prefix." + name1, 42);
+    node_local->declare_parameter("prefix." + name2, 100);
     // undeclare so that it doesn't interfere with the test
     node_local->undeclare_parameter("use_sim_time");
 
     {
       std::map<std::string, int64_t> actual;
-      EXPECT_TRUE(node_local->get_parameters("", actual));
+      EXPECT_TRUE(node_local->get_parameters("prefix", actual));
```
Comment (wjwwood) [inline rclcpp/test/rclcpp/test_node.cpp:1976]: Why were these changes needed?

File: rclcpp/test/rclcpp/test_qos_parameters.cpp [hunk 3]
```diff
@@ -234,7 +234,8 @@ TEST(TestQosParameters, declare_no_parameters_interface) {
 
   std::map<std::string, rclcpp::Parameter> qos_params;
   EXPECT_FALSE(
-    node->get_node_parameters_interface()->get_parameters_by_prefix("qos_overrides", qos_params));
+    node->get_node_parameters_interface()->get_parameters_by_prefix(
+      "qos_overrides./my/fully/qualified/topic_name", qos_params));
```
Comment (wjwwood) [inline rclcpp/test/rclcpp/test_qos_parameters.cpp:238]: I don't understand why this changed either (probably just missing something), but this might be an indicator that we're breaking behavior and therefore we need documentation of this, in the migration guide for galactic and/or user docs.

File: rclcpp/test/rclcpp/test_node.cpp (hunk 2 above)
Comment (ivanpauno) [reply rclcpp/test/rclcpp/test_node.cpp:1976]: because now there are more parameters automatically created when a node is created, and that made the test fail

File: rclcpp/test/rclcpp/test_qos_parameters.cpp (hunk 3 above)
Comment (ivanpauno) [reply rclcpp/test/rclcpp/test_qos_parameters.cpp:238]: Before, the only parameters that started with the prefix `qos_overrides` were the ones immediately declared above.
Now there's also the ones of the `/parameter_events` topic

File: rclcpp/include/rclcpp/node_interfaces/node_parameters_interface.hpp [hunk 4]
```diff
@@ -30,6 +30,18 @@
 
 namespace rclcpp
 {
+
+namespace detail
+{
+/// \internal Get the definitive parameter overrides.
+std::map<std::string, rclcpp::ParameterValue>
+resolve_parameter_overrides(
```
[the remark this answers is not in the document]
Comment (ivanpauno) [reply rclcpp/include/rclcpp/node_interfaces/node_parameters_interface.hpp]: Moved in https://github.com/ros2/rclcpp/pull/1532/commits/9449ef8b4986817c66950eb97027f5bebd2a6555

File: rclcpp/src/rclcpp/node.cpp (hunk 1 above)
Comment (ivanpauno) [reply rclcpp/src/rclcpp/node.cpp:177]: Done in https://github.com/ros2/rclcpp/pull/1532/commits/04ce4210e25566c694f6abbdd68586ed5905b4d2

**Comment by ivanpauno** [comment]
Rebased to solve conflicts

File: rclcpp/include/rclcpp/node_interfaces/node_parameters_interface.hpp [hunk 5]
```diff
@@ -30,6 +30,7 @@
 
 namespace rclcpp
 {
+
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/node_interfaces/node_parameters_interface.hpp:33]: nitpick: seems like an odd change for this pr

File: rclcpp/src/rclcpp/node.cpp [hunk 6]
```diff
@@ -146,6 +194,20 @@ Node::Node(
...
+  {
+    QosPolicyKind::Depth,
+    QosPolicyKind::Durability,
+    QosPolicyKind::History,
+    QosPolicyKind::Reliability,
+  },
+    node_parameters_,
+    node_topics_->resolve_topic_name("/parameter_events"),
+    options.parameter_event_qos(),
+    rclcpp::detail::PublisherQosParametersTraits{});
```
Comment (wjwwood) [inline rclcpp/src/rclcpp/node.cpp:210]: The indentation is weird here, is that uncrustify doing that?

I'd expect:

```suggestion
  rclcpp::detail::declare_qos_parameters(
    rclcpp::QosOverridingOptions{
      QosPolicyKind::Depth,
      QosPolicyKind::Durability,
      QosPolicyKind::History,
      QosPolicyKind::Reliability,
    },
    node_parameters_,
    node_topics_->resolve_topic_name("/parameter_events"),
    options.parameter_event_qos(),
    rclcpp::detail::PublisherQosParametersTraits{});
```

**Review by wjwwood — APPROVED** [review]
two small nitpicks but it looks good to me, could be merged now.

File: rclcpp/include/rclcpp/node_interfaces/node_parameters_interface.hpp (hunk 5 above)
Comment (ivanpauno) [reply rclcpp/include/rclcpp/node_interfaces/node_parameters_interface.hpp]: fixed in https://github.com/ros2/rclcpp/pull/1532/commits/124c948f28ecd09802900663c56dfe25f9370f6e

File: rclcpp/src/rclcpp/node.cpp (hunk 6 above)
Comment (ivanpauno) [reply rclcpp/src/rclcpp/node.cpp:210]: I agree with your suggestion, but this is a problem in our uncrustify settings.
I also tried:

```cpp
  rclcpp::detail::declare_qos_parameters(
    rclcpp::QosOverridingOptions({
      QosPolicyKind::Depth,
      QosPolicyKind::Durability,
      QosPolicyKind::History,
      QosPolicyKind::Reliability,
    }),
    ...
```

without success.

I preferred to follow uncrustify requested style, but I can also add a `INDENT-OFF` comment.

Comment (wjwwood) [reply rclcpp/src/rclcpp/node.cpp:210]: Nah it's fine as-is. Oh well 🤷‍♂️

**Review by wjwwood — APPROVED** [review]

**Comment by ivanpauno** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/13778/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8646/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11486/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/13855/)

**Comment by ivanpauno** [comment]
Force pushed after rebasing on master

**Comment by ivanpauno** [comment]
I think that the errors above were unrelated, but running CI again to compare with the latest nightlies:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/13861/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8716/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11570/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/13949/)

**Comment by ivanpauno** [comment]
CI after https://github.com/ros2/system_tests/pull/465:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/13879/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8733/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11588/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/13968/)

**Comment by ivanpauno** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/13924/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8767/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11628/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14010/)
